Add devc section

griffinhadfield opened the pull request:
Addressing https://github.com/OpenLiberty/guide-getting-started/issues/208

gkwan-ibm reviewed (changes requested):
@griffinhadfield 
- [x] improve the "Running the application in development mode with a container" section title
  - e.g. "Developing the application with Docker container"
- [x] liberty-maven-plugin in pom.xml is not the latest, so it cannot start `devc`
- [x] `guide-docker-dev-mode` is not right in the output for `docker ps`
- [x] when hit enter to test, test failed
- [ ] consider to add insturction to update the application with minor change to show the application can pick
  - a staging directory may need

AustinSeto reviewed (approved):
# Windows Peer Review

Should note that https://github.com/OpenLiberty/ci.maven/issues/1044 is still open and I ran into the same issue with this guide. Other than that, ran end-to-end and everything worked as expected. 

PR looks good to me

tt-le reviewed (approved):
Linux Peer Review
---

Ran end to end on linux and everything worked as expected.

README changes LGTM

dmuelle commented on README.adoc:
```diff
@@ -471,6 +471,79 @@ To remove the image, run the following command:
+The Open Liberty Maven plug-in includes a `devc` goal that simplifies developing
+your application in a Docker container. This goal builds a Docker
+image, mounts the required directories, binds the required ports, and then runs
+the application inside of a container. This development mode, known as dev mode, also listens for
```
Here,  dev mode in a container is presented as a separate dev mode ("this dev mode")  from what is discussed earlier in the guide- but is it really a different mode?  Wouldn't either example be an instance of the same "dev mode" functionality? There are two goals (dev vs devc), but they enable the same basic functions. In the docs, it is presented as "container support for dev mode" https://www.openliberty.io/docs/21.0.0.1/development-mode.html#_container_support_for_dev_mode. Similar for the maven doc: https://github.com/OpenLiberty/ci.maven/blob/master/docs/dev.md

I think this might be confusing for someone working through the entire guide. If dev mode, in and of itself, is different here than earlier in the guide, how and why? If it is not different, and is simply the same function being run by a different Maven goal, I think you could rewrite this to reflect that. Assume the reader has already been introduced to dev mode earlier in the guide, then explain how they can also run dev mode in a docker container. 


Furthermore, I'd recommend using "dev mode" throughout the guide, rather than "development mode". Alasdair asked that we make that change in the doc and it is consistent with documentation in the Git repos.

dmuelle commented on README.adoc:
````diff
@@ -471,6 +471,79 @@ To remove the image, run the following command:
+mvn liberty:devc
+```
+
+After you see the following message, your application server in dev mode is ready:
````
After you see the following message, your application server in dev mode is ready:
--->
When you see the following message, Open Liberty is ready to run in dev mode:

dmuelle commented on README.adoc:
````diff
@@ -471,6 +471,79 @@ To remove the image, run the following command:
+docker ps
+```
+
+Make sure that your container is running and does not have `Exited` as its status:
````
what should the user do if the status is "exited" ?

griffinhadfield replied:
I have replaced this sentence with `Your container should be running and have Up as its status:`. We don't usually present troubleshooting/error handling options in the guide, so I think this new sentence is more consistent with other guides.

dmuelle commented on README.adoc:
```diff
@@ -471,6 +471,79 @@ To remove the image, run the following command:
+
+To access the application, go to the http://localhost:9080/system/properties[http://localhost:9080/system/properties^] URL. 
+
+Similar to the development mode used previously in this guide, this
```
see previous comment re dev mode

dmuelle commented on README.adoc:
```diff
@@ -471,6 +471,79 @@ To remove the image, run the following command:
+----
+
+After you make the file changes, Open Liberty automatically reloads its
+configuration. You will now be able to access the application at the
```
You will now be able to access the application...
-->
You can access the application...

dmuelle commented on README.adoc:
```diff
@@ -471,6 +471,79 @@ To remove the image, run the following command:
+When you are finished, exit development mode by pressing `CTRL+C` in the
+command-line session that the container was started from, or by typing `q` and
+then pressing the `enter/return` key. This stops and removes the container for you. You
+can check that the container was stopped by running the `docker ps` command.
```
This stops and removes the container for you. You can check that the container was stopped by running the `docker ps` command.
--->
Either of these options stops and removes the container. To check that the container was stopped, run the `docker ps` command.

dmuelle reviewed (changes requested):
see my comments inline, lmk if any questions. Thanks

griffinhadfield commented:
@dmuelle Thank you very much for the feedback, I have addressed all of your points. The updated guide can be seen at http://lgdev1.fyre.ibm.com:4020/guides/getting-started.

Regarding replacing the term "development mode" with "dev mode" I have done so, but I have one question about the sentence that I introduced that term in: 
> Stop the Open Liberty server if it is running, and start it in development mode, known as dev mode, by running the `liberty:dev` goal in the `start` directory:

I added "known as dev mode" to this sentence, but I am wonder whether the sentence is now too clunky. What do you think? Should I instead keep the first mention of the mode there as "development mode", and add the ", known as dev mode," interjection to the next mention in the sentence after?

dmuelle commented:
@griffinhadfield I think you can remove the first mention as "development mode" and just use "dev mode" for all instances. That's consistent with other docs. Although we initially included "development mode" in the docs, it has since been updated per Alasdairs request to be just "dev mode"

griffinhadfield commented:
> @griffinhadfield I think you can remove the first mention as "development mode" and just use "dev mode" for all instances. That's consistent with other docs. Although we initially included "development mode" in the docs, it has since been updated per Alasdairs request to be just "dev mode"

@dmuelle Kk I have made that update and it, along with the other updates, can be seen here: http://lgdev1.fyre.ibm.com:4020/guides/getting-started

dmuelle commented on README.adoc:
```diff
@@ -323,8 +323,8 @@ You can also access the `/health/ready` endpoint by going to the http://localhos
 Making code changes and recompiling is fast and straightforward.
-The development mode of Open Liberty automatically picks up changes in the `.class` files and artifacts, without needing to be restarted.
-Alternatively, you can run the `run` goal and manually repackage or recompile the application by using the `mvn package` command or the `mvn compile` command while the server is running. The development mode was added to further improve the developer experience by minimizing turnaround times.
+The dev mode of Open Liberty automatically picks up changes in the `.class` files and artifacts, without needing to be restarted.
```
I'd recommend
The dev mode of Open Liberty
--->
Open Liberty dev mode

and

The dev mode was added...
-->
Dev mode was added..

dmuelle replied:
or, instead of "The dev mode of Open Liberty automatically picks up.."

"When Open Liberty runs in dev mode, it automatically picks up

dmuelle reviewed (approved):
LGTM thanks

yeekangc reviewed (approved):
Looking at the latest draft, LGTM.
